### PR TITLE
Use lighthouse node API and restructure output writing

### PR DESCRIPTION
**Pull request by calebeby**
# Overview

There is a lot of new code here! Happy to do a video walkthrough if that is helpful.

Here is what stayed mostly the same:
- The crawler code. In a follow-up PR I will probably swap out the crawler. For now, I just adapted the output so that it is an [async iterator](https://javascript.info/async-iterators-generators) instead of an event emitter. I moved `createUrlFilter` to a separate file without changing it.
- The CLI code/args parsing/progress displaying

Everything else was rewritten:
- `lighthouse.ts` now manages a pool of [`worker_threads`](https://nodejs.org/api/worker_threads.html) (Node's version of web workers). Each worker thread runs the `lighthouse-worker.ts` file. It figures out which worker is available to handle a lighthouse run, and tells it to run it.
- `lighthouse-worker.ts` launches a chrome instance, and runs lighthouse whenever it is requested to.
- `async-iterator-queue` is a data structure that has an async iterator that loops through its items, yielding as many as it has. It will pause when it runs out of items. As `push` is called more, the new items are yielded as well. When `finish` is called, the iterator exits. It is used to adapt the simplecrawler output to an async iterator for use in `main.ts`
- `main.ts` wires everything together. It is called from the `cli` code. It manages the state and pulls in the URLs from the crawler, requests a worker from `lighthouse.ts`, and sends the URL to the worker. Then it takes the result and updates the state, and passes the result to the `OutputWriter`.
- `output-writer/index.ts` contains the `OutputWriter` interface. Right now there is a CSV `OutputWriter` and a stub for a `google-sheets` `OutputWriter`. It also has a function to merge several `OutputWriter`s together so the code in `main` can pretend as though there is only one `OutputWriter` even when there are distinct outputs.

This PR targets `next`. Here are other upcoming changes I'm hoping to do in separate PR's before release:
- `google-sheets` output format
- Replace `simplecrawler`
- Add more flags for customizing lighthouse (and/or allow reading config file)
- Write documentation (in this PR I deleted partially-out-of-date documentation so I wouldn't forget to update it once all the CLI flags /etc are all settled. I'll copy it back and update it in a follow up PR).
- I would like to set up a test suite that runs a static HTML server and runs lighthouse-parade against it, which we can use as a non-changing reference, which will be helpful to review lighthouse updates, etc.
- Bundle the JS output instead of directly using output from `tsc`
- I will write changesets.

# For review

I would suggest turning off whitespace changes, which will help just a little bit.
Also GitHub has the nice new file tree on the left of the review page, it is probably helpful to use that.

# Testing

Create a baseline lighthouse run to compare against:
- `git checkout main`
- `npm i`
- `npm run build`
- `npm i -g lighthouse-parade`
- `npm link`
- `lighthouse-parade ...`

Try the new changes
- `git checkout lighthouse-node-api`
- `npm i`
- `npm run build`
- `npm link`
- `lighthouse-parade ...`

Then you can compare the generated CSV data and make sure the differences are minor. The new CSV format won't work with the existing google sheets template, so you can just review the CSV directly or in Numbers or something for now.

### CODE REVIEW

File: .changeset/curly-paws-melt.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'lighthouse-parade': major
+---
+
+Rename `--ignore-robots` to `--ignore-robots-txt`
```
Comment (gerardo-rodriguez) [inline .changeset/curly-paws-melt.md:5]: I expected to see a different description here. 🤔

File: README.md [hunk 2]
```diff
@@ -10,58 +10,19 @@ There are great tools for doing performance analysis on a single web page. We us
...
 ## How?
 
-With a single command, the tool will crawl an entire site, run a Lighthouse report for each page, and then output a spreadsheet with the aggregated data. There is an [example of the generated spreadsheet](./examples/exampleAggregatedMobileReport.csv) in the `examples/` directory of this project. Each row in the spreadsheet is a page on the site, and each individual performance metric is a column. This makes it very easy to perform high-level analysis because you can sort the rows by whichever metric you are analyzing. This immediately shows the best and worst performing pages. In the following example, the rows are sorted by first contentful paint.
-
-![Lighthouse data from multiple reports aggregated into a single spreadsheet](./assets/lighthouse-data-by-lcp.png)
-
-It is also easy to graph data in this format. The following example is a histogram showing a problematic distribution of largest contentful paint scores.
-
-![Histogram showing LCP scores](./assets/lcp_histogram.svg)
+TODO: rewrite/see what to keep
```
[the remark this answers is not in the document]
Comment (calebeby) [reply README.md:13]: Yep

File: src/output-writer/csv-writer.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  let mutexPromise: Promise<unknown> = Promise.resolve();
+  const columns: CSVColumn[] = [];
+  return {
+    async addEntry(report) {
+      mutexPromise = mutexPromise.then(async () => {
+        if (!hasWrittenHeader) {
+          hasWrittenHeader = true;
+          for (const category of Object.values(report.categories)) {
+            columns.push({
+              name: `${category.title} (overall score)`,
```
Comment (gerardo-rodriguez) [inline src/output-writer/csv-writer.ts:51]: Not a change request for this PR, but seeing "overall score" hardcoded in English made me wonder if the tool could support different language outputs in the future. 🙂

File: src/main.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  );
+
+  return combineOutputWriters(outputWriters);
+};
+
+export const enum State {
+  Pending,
+  ReportInProgress,
+  ReportSuccess,
+  ReportFailure,
```
Comment (gerardo-rodriguez) [inline src/main.ts:37]: Should this be `ReportPending` to match the format of the others?
```suggestion
  ReportPending,
  ReportInProgress,
  ReportSuccess,
  ReportFailure,
```

Comment (calebeby) [reply src/main.ts:37]: Maybe I should just take Report off of it... or change it to Lighthouse. The bottom three represent the state of lighthouse, but the first one means the URL hasn't been given to lighthouse yet.

Comment (gerardo-rodriguez) [reply src/main.ts:37]: Ah, I see. Hmm...maybe remove "Report" and leave them as `Pending`, `InProgress`, `Success`, `Failure` then as you suggested is fine.

File: src/output-writer/csv-writer.ts (hunk 3 above)
Comment (calebeby) [reply src/output-writer/csv-writer.ts:51]: fun to read these in Spanish: https://github.com/microsoft/TypeScript/blob/main/src/loc/lcl/esn/diagnosticMessages/diagnosticMessages.generated.json.lcl#L463

File: src/async-iterator-queue.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        await new Promise<void>((resolve) => {
+          unpauseGenerator = resolve;
+        });
+      }
+    },
+  };
+};
+
+if (import.meta.vitest) {
+  const { it, expect } = import.meta.vitest;
```
[the remark this answers is not in the document]
Comment (gerardo-rodriguez) [reply src/async-iterator-queue.ts:55]: LOL

FWIW, I do like the `it` "does a thing" format.

Nothing actionable, I was only curious. 🙃

**Review by gerardo-rodriguez — APPROVED** [review]
Pending [one suggestion](https://github.com/cloudfour/lighthouse-parade/pull/111#discussion_r905295263) LGTM, great work, @calebeby! 🎉

**Pull request merged**